js/modules/k6/crypto/x509/x509.go

```diff
@@ -0,0 +1,220 @@
+	StateOrProvinceName string `js:"stateOrProvinceName"`
+	LocalityName        string `js:"localityName"`
+	OrganizationName    string `js:"organizationName"`
+}
+
+// PublicKey is a public key
+type PublicKey struct {
+	Algorithm string
+	E         int
+	N         []byte
+}
+
+// New constructs the X509 interface
+func New() *X509 {
+	return &X509{}
+}
+
+// Parse produces an entire x.509 certificate
+func (X509) Parse(ctx context.Context, encoded string) Certificate {
+	parsed := parseCertificate(ctx, encoded)
+	return makeCertificate(parsed)
+}
+
+// GetAltNames extracts alt names
+func (X509) GetAltNames(ctx context.Context, encoded string) []string {
+	parsed := parseCertificate(ctx, encoded)
+	return altNames(parsed)
+}
+
+// GetIssuer extracts certificate issuer
+func (X509) GetIssuer(ctx context.Context, encoded string) Issuer {
+	parsed := parseCertificate(ctx, encoded)
+	return makeIssuer(parsed.Issuer)
+}
+
+// GetSubject extracts certificate subject
+func (X509) GetSubject(ctx context.Context, encoded string) Subject {
+	parsed := parseCertificate(ctx, encoded)
+	return makeSubject(parsed.Subject)
+}
+
+func parseCertificate(ctx context.Context, encoded string) *x509.Certificate {
+	decoded, _ := pem.Decode([]byte(encoded))
+	if decoded == nil {
+		err := errors.New("failed to decode certificate PEM file")
+		common.Throw(common.GetRuntime(ctx), err)
+	}
+	parsed, err := x509.ParseCertificate(decoded.Bytes)
+	if err != nil {
+		err := errors.New("failed to parse certificate")
+		common.Throw(common.GetRuntime(ctx), err)
+	}
+	return parsed
+}
+
+func makeCertificate(parsed *x509.Certificate) Certificate {
+	return Certificate{
+		Subject:            makeSubject(parsed.Subject),
+		Issuer:             makeIssuer(parsed.Issuer),
+		NotBefore:          iso8601(parsed.NotBefore),
+		NotAfter:           iso8601(parsed.NotAfter),
+		AltNames:           altNames(parsed),
+		SignatureAlgorithm: signatureAlgorithm(parsed.SignatureAlgorithm),
+		FingerPrint:        fingerPrint(parsed),
+		PublicKey:          makePublicKey(parsed),
+	}
+}
+
+func makeSubject(subject pkix.Name) Subject {
+	return Subject{
+		CommonName:             subject.CommonName,
+		Country:                first(subject.Country),
+		PostalCode:             first(subject.PostalCode),
+		StateOrProvinceName:    first(subject.Province),
+		LocalityName:           first(subject.Locality),
+		StreetAddress:          first(subject.StreetAddress),
+		OrganizationName:       first(subject.Organization),
+		OrganizationalUnitName: subject.OrganizationalUnit,
+	}
+}
+
+func makeIssuer(issuer pkix.Name) Issuer {
+	return Issuer{
+		CommonName:          issuer.CommonName,
+		Country:             first(issuer.Country),
+		StateOrProvinceName: first(issuer.Province),
+		LocalityName:        first(issuer.Locality),
+		OrganizationName:    first(issuer.Organization),
+	}
+}
+
+func makePublicKey(parsed *x509.Certificate) PublicKey {
+	key := parsed.PublicKey.(*rsa.PublicKey)
```

Comment:
This type assertion should probably be checked like [this](https://tour.golang.org/methods/15) or [this](https://tour.golang.org/methods/16) and a normal error returned for unsupported key types, since currently it will panic when the public key isn't RSA.

Reply:
Started making this change, but there might be an improvement we can make here.

In the signing code I've ended up with a combined `x509.PublicKey` type that can hold a key from any of the 3 cryptosystems, RSA DSA ECDSA. What about using that type in the `publicKey` section here?

The verify API seems to want the ability to verify with `certificate.publicKey`. This would make it possible.

We'd have to move `certificate.publicKey.algorithm` out to eg `certificate.publicKeyAlgorithm`. That matches how the golang `Certificate` type does it.

Reply:
This seems very sensible to me :+1: